docs/blueprints/EP029.rst

```diff
@@ -0,0 +1,103 @@
+ - Unavailability: unavailable resources are the ones that were administratively configured not to be used (e.g., due to a maintenance window, traffic engineering, etc.)
+
+Path Disjointness
+----------------
+
+Paths/routes between a given pair of source and destination in a network are called link disjoint if they have no links in common and node disjoint if they have no common nodes besides the source and destination nodes. Disjoint paths are required for reliability in communication once one can be used as a protection/failover route of the other. Ideally, the maximum disjoints an alternative route, the better (node and link disjoint). However, considering practical scenarios and limitations on the network topology, finding perfectly disjoint paths can become a complex task (especially if we also consider additional user requirements for the path, such as maximum end-to-end latency, for instance). Thus, it is reasonable to assume some hypotheses: i) nodes (switches) usually have more protection than links and are less susceptible to failures, so that link-disjoint usually offers a good level of protection; ii) in a medium/large topology, finding the entirely disjoint path can be very time consuming so that finding the maximum disjoint path among a few alternatives can be a more efficient approach.
+
+For this document, we define the disjointness level of a path as the complementary percentage of shared links with another path. For example, consider the following topology and let P1(L1,L2,L3) be a path between source node S and destination node D, composed of the links L1, L2, and L3. Furthermore, let P2(L4,L5,L3) and P3(L4,L5,L6,L7) be two alternative paths between S and D. The disjointness level between P1 and P2 is 66.6% (1 - 1/3), whereas the disjointness level between P1 and P3 is 100% (1 - 0/3). In this scenario, P3 is a better disjoint path from P1 than P2 is from P1. Note that there is another path P4(L4,L5,L6,L8,L9), that offers the same disjointness level as P3; however, P3 has the best cost than P4 (e.g., lower hop count).
+
+  .. image:: https://drive.google.com/uc?export=view&id=17kdq-S9ZZTySGsUaIYIWDxucUVSY0c7a
+
+Path Failover Approach
+======================
+
+To provide service protection for EVCs, MEF E-Line will use the following strategy:
+
+ - EVCs will have their primary path configured according to the user requirements (static or dynamic paths) and a failover/protection path. The failover/protection path will have its flows installed in advance (before failure events), including the forwarding rules for all switches in the path and egress forwarding rules at the UNI switches. Thus, only the ingress forwarding rules will have to be configured during a link failure.
```

Comment:
This has probably been tested but what would happen when an UNI doesn't have multiple uplinks? Such as, this path between A and Z:

````
A --- B --- C --- D --- Z
      \---E---F--/
````

Assuming the primary path is A-B-C-D-Z, would it leave pre-provisioned E-F-D-Z? And then when the failure happens on provision the remaining B-E segment? Or would it still doesn't matter and then it would pre-provision the entire B-E-F-D-Z and then only provision A-B when the it would need to switchover?

Reply:
Thank you for the question, Vinicius, and for the opportunity to discuss some possible strategies for this case! Some notes about your question:

1. When the UNI doesn't have multiple uplinks, the maximum disjoint path algorithm will try to find the "best available" alternative that provides a backup path with a higher disjointness degree. In other words, even though we know that when a UNI does not have multiple uplinks we actually have a risky situation (because failures on the first hop would impact that EVC), the Service Protection blueprint tries to maximize the protection of this EVC by choosing alternative paths that can still provide some backup for the EVC.
2. For the example provided, and considering the primary path is A-B-C-D-Z, we actually would leave pre-provisioned A-B-E-F-D-Z. It is worth emphasizing that the links A-B and D-Z will use different service tags for this new protection/failover path.
3. When a failure happens in the above scenario, the failover strategy should only reconfigure the ingress OF rules on A and Z.
4. No need to provisioning A-B during a failure, it will be pre-provisioned earlier (using different service tags). During a failure, only the ingress OF rules will be setup on A and Z.

I hope the above comments can help make this more clear, but if not please let me know!

Reply:
@italovalcy got it, appreciated this information, thanks. A have a follow up question. Let's assume this example uses an EPL. If the protection path is fully pre-provisioned, are we always setting a lower priority for the protection path in the ingress link just so it's not matched? and then when the time comes to switchover it removes the higher priority entry of the active path of the EVC?

Reply:
Hi @viniarck, thanks for the additional question! Yes, the protection/failover path is fully pre-provisioned except for the ingress rules. Having the ingress rules as lower-priority was an option but it would help much because during a failure we would have to send two FlowMods anyway. The idea was then to leave the ingress flows to be installed during the failure (also two FlowMods).

In the example above, we would have the following flows then (I've also included h1 and h2 to represent the User, which will result in the flows for the actual UNIs):

1. Current path:
```
h1 -> A -> B -> C -> D -> Z -> h2
h2 -> Z -> D -> C -> B -> A -> h1
```
2. Failover path:
```
A -> B -> E -> F -> D -> Z -> h2
Z -> D -> F -> E -> B -> A -> h1
```

Then, during a failure, we have to send two FlowMods, particularly: `h1 -> A` and `h2 -> Z`. It is worth mentioning that the above setup does not interfere with provisioned user traffic (either for EPL or EVPL).

I'll send a commit with those examples in the blueprint since I consider these questions very relevant and examples can be helpful to understand the strategy.

Reply:
Commit [ea55f22](https://github.com/kytos-ng/kytos/commit/ea55f2282cda9205e09fe8687e9c28b8f046d598) was added to cover the examples discussed here. @viniarck please let me know if you have additional comments

Reply:
Awesome, thanks @italovalcy, I don't have more questions, let me close this thead.